File: algorithms/sort/python/quick_sort.py
from typing import Optional

import numpy as np


class QuickSort:

    def __init__(self, array: list | np.ndarray) -> None:
        self.array = array

    def partition(self, left: int, right: int) -> int:
        pivot_value = self.array[right]
        i: int = left - 1
        for j in range(left, right):
            if self.array[j] <= pivot_value:
                i += 1
                self.array[i], self.array[j] = self.array[j], self.array[i]
        self.array[i + 1], self.array[right] = self.array[right], self.array[i + 1]
        return i + 1

    def quick_sort(
        self, left: Optional[int] = None, right: Optional[int] = None
    ) -> list | np.ndarray:
        if left is None:
            left = 0
        if right is None:
            right = len(self.array) - 1
        if left < right:
            pivot_index = self.partition(left, right)
            self.quick_sort(left, pivot_index - 1)
            self.quick_sort(pivot_index + 1, right)

        return self.array


if __name__ == "__main__":
    array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    np.random.shuffle(array)
    qs = QuickSort(array)
    print("-" * 20, " array ", "-" * 20)
    print(qs.array)
    print("-" * 20, " sorted arrray ", "-" * 20)
    print(qs.quick_sort())

File: algorithms/sort/c++/quick_sort.cpp
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

int partition(vector<int> &array, int left, int right) {
  int pivot_value = array[right];
  int i = left - 1;
  int j;
  for (j = left; j < right; j++) {
    if (array[j] <= pivot_value) {
      i = i + 1;
      swap(array[i], array[j]);
    }
  }
  swap(array[i + 1], array[right]);
  return i + 1;
}

void quickSort(vector<int> &array, int left, int right) {
  if (left < right) {
    int pivot_index = partition(array, left, right);
    quickSort(array, left, pivot_index - 1);
    quickSort(array, pivot_index + 1, right);
  }
}

void printVector(const vector<int> &arr) {
  for (int val : arr) {
    cout << val << " ";
  }
  cout << endl;
}

int main() {
  vector<int> array = {4, 1, 4, 6, 2, 4, 7, 8, 3, 4, 2, 2, 6, 7, 9, 4};
  int left = 0;
  int right = array.size() - 1;

  quickSort(array, left, right);
  printVector(array);

  return 0;
}